syntax = "proto3";

package distributedstore;

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);

  rpc get(GetRequest) returns (GetResponse);

  rpc slowDown(SlowdownRequest) returns (SlowDownResponse);

  rpc restore(RestoreRequest) returns (RestoreResponse);

  rpc canCommit(CommitRequest) returns (CommitRespone);

  rpc doCommit(doCommitRequest) returns (doCommitRespone);
  
  rpc discoverMessage(dMessage) returns (Empty);

  rpc askVotePut(askVotePutRequest) returns (askVotePutRespone);

  rpc askVoteGet(askVoteGetRequest) returns (askVoteGetRespone);
}

message askVoteGetRequest{
  string key = 1;
}

message askVoteGetRespone{
  bool success = 1;
  string value = 2;
  int32 vote_size = 3;
}

message askVotePutRespone{
  bool success = 1;
  int32 vote_size = 2;
}

message askVotePutRequest{
  string key = 1;
  string value = 2;
}

message dMessage{
  string ip = 1;
  int32 port = 2;
}

message doCommitRespone{
   bool success = 1;
}

message doCommitRequest{
  string key = 1;
  string value = 2;
}

message CommitRequest{
  string key = 1;
  string value = 2;
}

message CommitRespone {
  bool success = 1;
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
}

message SlowdownRequest {
  int32 delay = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message RestoreRequest {}

message RestoreResponse {
  bool success = 1;
}

message Empty {
    
}